src/main/java/algorithms/sorting/radixSort/RadixSort.java

```diff
@@ -1,53 +1,69 @@
-        return (num & mask) >> (segment * NUM_BITS);
-    }
+  private static final int NUM_BITS = 8;
+  private static final int NUM_SEGMENTS = 4;
 
-    private static void radixSort(int[] arr, int[] sorted) {
-        // sort the N numbers by segments, starting from left-most segment
-        for (int i = 0; i < NUM_SEGMENTS; i++) { 
-            int[] freqMap = new int[1 << NUM_BITS]; // at most this number of elements
+  /**
+   * Creates masking on the segment to obtain the value of the digit.
+   *
+   * @param num The number.
+   * @param segment The segment we are interested in.
+   *
+   * @return The value of the digit in the number at the given segment.
+   */
+  private static int getSegmentMasked(int num, int segment) {
+    int mask = ((1 << NUM_BITS) - 1) << (segment * NUM_BITS);
+    return (num & mask) >> (segment * NUM_BITS);
+  }
 
-            // count each element
-            for (int num : arr) {
-                freqMap[getSegmentMasked(num, i)]++;
-            }
-            // get prefix sum
-            for (int j = 1; j < freqMap.length; j++) {
-                freqMap[j] += freqMap[j-1];
-            }
-            // place each number in its correct sorted position up until the given segment
-            for (int k = arr.length-1; k >= 0; k--) {
-                int curr = arr[k];
-                int id = getSegmentMasked(curr, i);
-                sorted[freqMap[id] - 1] = curr;
-                freqMap[id]--;
-            }
-            // we do a swap so that our results above for this segment is saved and passed as input to the next segment
-            int[] tmp = arr;
-            arr = sorted;
-            sorted = tmp;
-        }
-        sorted = arr;
-    }
+  /**
+   * Radix sorts a given input array and updates the output array in-place.
+   *
+   * @param arr original input array.
+   * @param sorted output array.
+   */
+  private static void radixSort(int[] arr, int[] sorted) {
+    // sort the N numbers by segments, starting from left-most segment
+    for (int i = 0; i < NUM_SEGMENTS; i++) {
+      int[] freqMap = new int[1 << NUM_BITS]; // at most this number of elements
 
-    public static void radixSort(int[] arr) {
-        int[] sorted = new int[arr.length];
-        radixSort(arr, sorted);
-        arr = sorted; // swap back lol
+      // count each element
+      for (int num : arr) {
+        freqMap[getSegmentMasked(num, i)]++;
+      }
+      // get prefix sum
+      for (int j = 1; j < freqMap.length; j++) {
+        freqMap[j] += freqMap[j - 1];
+      }
+      // place each number in its correct sorted position up until the given segment
+      for (int k = arr.length - 1; k >= 0; k--) {
+        int curr = arr[k];
+        int id = getSegmentMasked(curr, i);
+        sorted[freqMap[id] - 1] = curr;
+        freqMap[id]--;
+      }
+      // we do a swap so that our results above for this segment is
+      // saved and passed as input to the next segment
+      int[] tmp = arr;
```

Comment:
we should mention this is a implementation trick. allowing O(n) space since we are re-using the arrays, one to save our data and another to iteratively work towards the result using the next segment. Swapping back and forth the two as reuse

Reply:
Added more details

Reply:
yup, in general, one can implement a simpler
O((b/r) * (2^r+n)) space complexity
where we have b/r passes and each pass of counting sort takes (k+n) space where k here is 2^r (range) (assuming we initialize a new array at each pass)